pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,792 @@
+	};
+	use types::{
+		ActiveLoan, AssetOf, BorrowLoanError, CloseLoanError, ClosedLoan, CreateLoanError,
+		CreatedLoan, LoanInfoOf, PortfolioValuation, PortfolioValuationUpdateType, WriteOffState,
+		WriteOffStatus, WrittenOffError,
+	};
+
+	use super::*;
+
+	type PoolIdOf<T> = <<T as Config>::Pool as PoolInspect<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::CurrencyId,
+	>>::PoolId;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify a curreny.
+		type CurrencyId: Parameter + Copy + MaxEncodedLen;
+
+		/// Identify a non fungible collection
+		type CollectionId: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ TypeInfo
+			+ Copy
+			+ MaxEncodedLen;
+
+		/// Identify a non fungible item
+		type ItemId: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ TypeInfo
+			+ Copy
+			+ MaxEncodedLen;
+
+		/// Identify a loan in the pallet
+		type LoanId: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ TypeInfo
+			+ MaxEncodedLen
+			+ Copy
+			+ EnsureAdd
+			+ One;
+
+		/// Defines the rate type used for math computations
+		type Rate: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ FixedPointNumber
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// Defines the balance type used for math computations
+		type Balance: tokens::Balance + FixedPointOperand;
+
+		/// Fetching method for the time of the current block
+		type Time: UnixTime;
+
+		/// Used to mint, transfer, and inspect assets.
+		type NonFungible: Transfer<Self::AccountId>
+			+ Inspect<Self::AccountId, CollectionId = Self::CollectionId, ItemId = Self::ItemId>;
+
+		/// Access to the pool
+		type Pool: PoolReserve<Self::AccountId, Self::CurrencyId, Balance = Self::Balance>;
+
+		/// Used to verify permissions of users
+		type Permissions: Permissions<
+			Self::AccountId,
+			Scope = PermissionScope<PoolIdOf<Self>, Self::CurrencyId>,
+			Role = Role,
+			Error = DispatchError,
+		>;
+
+		/// Used to calculate interest accrual for debt.
+		type InterestAccrual: InterestAccrual<
+			Self::Rate,
+			Self::Balance,
+			Adjustment<Self::Balance>,
+			NormalizedDebt = Self::Balance,
+		>;
+
+		/// Max number of active loans per pool.
+		#[pallet::constant]
+		type MaxActiveLoansPerPool: Get<u32>;
+
+		/// Max number of write-off groups per pool.
+		#[pallet::constant]
+		type MaxWriteOffGroups: Get<u32>;
+	}
+
+	/// Contains the last loan id generated
+	#[pallet::storage]
+	pub(crate) type LastLoanId<T: Config> =
+		StorageMap<_, Blake2_128Concat, PoolIdOf<T>, T::LoanId, ValueQuery>;
+
+	/// Storage for loans that has been created but are not still active.
+	#[pallet::storage]
+	pub(crate) type CreatedLoans<T: Config> = StorageDoubleMap<
```

Comment:
Do you mean the name? It contains several loans, and we already have an entity called `CreateLoan`. Why singular?

Reply:
But you query for a single loan. Currently this is mostly convention. But maybe it isn't. Just was my understanding

Reply:
For me, the type shows what it represents (a bunch of elements), and the query is the action over it. Also, a `CreatedLoan` type is iterable, which becomes more evident if the name is `CreatedLoans`. At least, this is how I see it 😅

Reply:
To give another reason here, the number of loans you could have in `CreatedLoans` is the same as `ActiveLoans`; the difference is the first is in a double map, and the second one map + vector. But both types are isomorphic, so they should be called the same way.

Reply:
If I read the code correctly they are especially not Isomorphic but disjoint sets. 
E.g. 
```
A is the set of all loans for a pool
B is the set of all created loans
C is the set of all active loans
D is the set of all closed loans

A = B v C v D;

B ^ C = {};
```

See: `CreatedLoans::<T>::take(..)`.

And arguably all names refer to the final value queried: 
* a single loan for `CreatedLoans`
* a vec of loans in `ActiveLoans`

Reply:
By isomorphic, I mean they have the same structure. Both types contain a lot of loans per pool; they are containers in the same way, you only change how you store them because of performance reasons, but they are able to store the "same" amount of things. That's why I like to put that `s`. 

It's true that if you call the type by how you get the value by the `get`/`take` method, then you should call it as you say, but it's the first time I see that point of view.

Looking at the system pallet, it's true that it seems to follow that agreement, so I'll change these names! (although I don't fully understand the reason 🤔). Thanks for pointing this out, I've never realized it!